File: examples/Test-servocontroller-2.cpp
/*------------------------------------------------------------------------*/
/* Test-servocontroller2                                                  */
/*------------------------------------------------------------------------*/
/* (c) Juan Gonzalez. July-2010                                           */
/* Updated by David Estevez-Fernandez, 2013                               */
/*------------------------------------------------------------------------*/
/* GPL license                                                            */
/*------------------------------------------------------------------------*/
/* Testing the Servocontroller.                                           */
/* This example sets the position of the servos 0 and 1 to 45 and -45      */
/* alternately                                                            */
/**************************************************************************/

#include "TestBase.h"

class Example : public TestBase
{
  public:
    Example(string envfile,string controller,bool showgui=true) :
       TestBase(envfile,controller,showgui) {};
    void run(dReal step, bool realtime=true);
};

void Example::run(dReal step, bool realtime)
{

  char key;
  std::cout << "Press a key to start the simulation" << std::endl;
  std::cin.get(key);

  penv->StartSimulation(step, realtime);
  stringstream is, os;

  //-- Main loop
  while(1) {

    //-- Servo positioning example 1: Using the setpos command. The position
    //-- of the two servos is given: Servo 0 to 45 and servo 1 to -45
    is << "setpos 45 -45 ";
    pcontroller->SendCommand(os,is);
    cout << "works ok";
    //-- Wait one second
    sleep(1);

    //-- Servo positioning example 2:
    //-- One method: using the "setpos" command:
    //-- is << "setpos -45 45 ";
    //-- pcontroller->SendCommand(os,is);

    //-- Another method: using the setpos1 command:
    //-- Seting the servo 0 position:
    is << "setpos1 0 -45 ";
    pcontroller->SendCommand(os,is);

    //-- Seting the servo 1 position:
    is << "setpos1 1 45 ";
    pcontroller->SendCommand(os,is);

    //-- Wait one second
    sleep(1);

    //-- Debug! Show the viewer transformation
    //RaveTransform<float> t = viewer->GetCameraTransform();
    //cout << "Transform: " << t << endl;
  }

}

int main(int argc, char ** argv)
{
    std::string envfile;

    if (argc==1)
      //-- Default file
      envfile="../models/Unimod2.env.xml";
    else
      envfile = argv[1];

  Example example(envfile, "servocontroller");

  //example.SetCamera(-0.256, -0.194, 0.54, 0.778, -0.155, 0.23, 0.119);
  example.run(0.005);

  return 0;
}

File: examples/Test-servocontroller-2.py
#!/usr/bin/env python
#// This program is free software: you can redistribute it and/or modify
#// it under the terms of the GNU Lesser General Public License as published by
#// the Free Software Foundation, either version 3 of the License, or
#// at your option) any later version.
#//
#// This program is distributed in the hope that it will be useful,
#// but WITHOUT ANY WARRANTY; without even the implied warranty of
#// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#// GNU Lesser General Public License for more details.
#//
#// You should have received a copy of the GNU Lesser General Public License
#// along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import with_statement # for python 2.5
__author__ = 'Juan Gonzalez'
__copyright__ = 'Copyright (C) 2010 Juan Gonzalez (juan@iearobotics.com)'
__license__ = 'GPLv3 license'

#-- This is an example for testing the servocontroller located in the
#-- Modular Robot Open Rave plugin
#-- The two servos of the Minicube-I modular robot are both set to 45
#-- and -45 degrees.
#-- The Minicube-I modular robot is composed of two Y1 modules.


from openravepy import *
from numpy import *
import time
import sys

def trans(T,x,y,z):
    T[0,3]=x; T[1,3]=y; T[2,3]=z;
    return T

def run():

    #-- Read the name of the xml fiel passed as an argument
    #-- or use the default name
    try:
        file_env = sys.argv[1]
    except IndexError:
        file_env = 'models/Unimod2.env.xml'

    env = Environment()
    env.Load(file_env)
    env.SetViewer('qtcoin')

    #-- Configure the camera view
    #-- Rotation
    T = matrixFromQuat(array((0.505073, 0.268078, 0.395983, 0.718493)))

    #-- Translation
    T=trans(T,0.412915, 0.156822, 0.285362)

    env.SetCamera(T)

    #-- Set the robot controller and start the simulation
    with env:
        robot = env.GetRobots()[0]
        robot.SetController(env.CreateController('servocontroller'))
        env.StopSimulation()
        env.StartSimulation(timestep=0.001)

    starttime = time.time()

    #-- In this example, it does not matter the number of servos of the robot
    #-- All of them are moved to 45 and -45 degrees alternately
    while True:
        #-- Set the servo position to 45 degrees for all the servos
        ref_pos = [45 for j in robot.GetJoints()]
        robot.GetController().SendCommand('setpos '+' '.join(str(f) for f in ref_pos))
        time.sleep(1.0)


        #-- Set the servo position to -45 degrees for all the servos
        ref_pos = [-45 for j in robot.GetJoints()]
        robot.GetController().SendCommand('setpos '+' '.join(str(f) for f in ref_pos))
        time.sleep(1.0)

if __name__=='__main__':
    run()
